Java plugin process

kodebach opened the pull request:
## Changes

Adds a new plugin `stdioproc` that can be used to put plugins into a separate process (an alternative to `process`). Also adds first implementation in the JNA binding.

Closes #3891

## Basics

- [x] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [x] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [x] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [x] The PR is rebased with current master.

## Checklist

- [x] I added unit tests for my code
- [x] I fully described what my PR does in the documentation
      (not in the PR description)
- [x] I fixed all affected documentation
- [x] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [x] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [x] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

kodebach commented:
@markus2330 I had to create a separate plugin, because `process` and the `pluginprocess` library make various assumptions that cannot easily be met from the Java side. The new `stdioproc` only needs an executable that observes a certain protocol on stdin/stdout. The unit tests show that this could even be achieved via a shell script (although I didn't implement proper parsing of `dump` in shell code).

I also re-implemented the `dump` format in Java. This may not be entirely required (in theory we could call the native code), but this variant is much easier to debug.

markus2330 commented:
> I had to create a separate plugin, because process and the pluginprocess library make various assumptions that cannot easily be met from the Java side.

Can this new plugin be used to replace the old? Afaik the old one wasn't used anyway.

> The new stdioproc

Library or plugin? Maybe we find a better name.

kodebach commented:
> Can this new plugin be used to replace the old? Afaik the old one wasn't used anyway.

The new plugin is more general, than the old one. It can work with any executable. For example, the `testmod` tests use a simple shell script. If this new approach works well (so far it looks good), we can probably delete the old `jni` plugin.

For the old `process` plugin, things are bit different. It just calls and existing plugin (via the C API) in another process. I have no idea why that could be useful, but if it is needed, the new `stdioproc` cannot be used to replace it directly. To replace `process` with `stdioproc` we would need a simple executable that calls a plugin via the C API.

To replace the `python` plugin (also uses the `pluginprocess` library) with `stdioproc`, we could just write a small Python script similar to the Java app in this PR.

> Library or plugin? Maybe we find a better name.

Plugin. Suggestions for a better name are welcome. The current name comes from "Standard I/O" (or `stdin`/`stdout`) and (external) process.

markus2330 commented on src/plugins/stdioproc/README.md:
````diff
@@ -0,0 +1,175 @@
+Set the config key `app` and the arrays `args/#`, `env/#` to the path of an executable, the arguments that shall be passed and the environment variables to be set.
+
+```
+kdb mount test.dump /tests/stdioproc specload 'app=/usr/bin/pluginproc' 'args=#1' 'args/#0=--load-plugin' 'args/#1=myplugin'
````
Is this ready to review? Is this really specload?

kodebach replied:
> Is this ready to review?

There aren't any tests which check that running a Java Plugin via `stdioproc` actually works. I've checked it manually and I'm currently working on automatic tests. Also some parts of the docs still need updates. Otherwise this PR should be done.

> Is this really specload?
It is somewhat similar, but
1. serves an entirely different purpose
2. `specload` uses only one-way communication, while `stdioproc` is two-way
3. `specload` only sends a single keyset, `stdioproc` has a more complex protocol

kodebach replied:
> Is this really specload?

I missed that the line literally says "specload". That's a copy paste error. It's fixed in the next commit.

markus2330 commented:
Please put the API changes to another branch, we do not have the time to review and discuss all of this. (@tucek also needs to be involved.) Please focus on allowing JNI plugins to run as robust as possible. We want to release this in February and there should be as little experimental code as possible.

kodebach commented:
> Please put the API changes to another branch

I can reduce the API changes, but it won't work with zero changes to the `Plugin` interface. In fact, I'm not sure how the `Plugin` interface was ever supposed to work. The `Plugin.getConfig()` method is just wrong. It makes no sense for Plugin to implement it's own `getConfig` method. The `config` keyset is passed to `Plugin.open()` and if it is needed outside this method it should be stored. However, a `getConfig()` method that could return an arbitrary `KeySet` (completely different from the user defined config) makes no sense.

> Please focus on allowing JNI plugins to run as robust as possible.

It should be pretty robust already. I tried to keep the whole protocol as simple as possible. That's also why I used the `dump` format. As long as no binary keys are involved, the whole protocol is human readable and therefore easy to debug.

> We want to release this in February and there should be as little experimental code as possible.

I assume you are talking about the `StdioProcPlugin` class. I will move changes related to this class to a separate PR, but I really think we should merge this before CM as well. The new API is much more Java-esque. In particular, it hides the fact that we use `Key`s to store errors, which is a very weird concept in Java. At the very least we need something like the `ErrorWriter` class, so that errors are added correctly. The old `WhitelistPlugin` only added `number` and `reason` for errors and warnings.

I also want to emphasize that I didn't actually change any existing APIs. I only added new APIs. The old JNI interface is still there.

kodebach commented:
@markus2330 I have now rewritten this PR to use the existing Java `Plugin` interface with as few changes as possible. The one remaining change is that I removed the broken `getConfig()`. I also didn't add `ErrorWriter` after all. The `Key` class already had `setError`/`addWarning` methods that were almost correct. I updated those instead.

Apart from tests/docs the PR is now fully read for review.

kodebach commented on src/bindings/jna/stdioproc/src/main/java/org/libelektra/stdioproc/StdIoProc.java:
```diff
@@ -0,0 +1,194 @@
+    println("ELEKTRA_STDIOPROC ACK v1");
+    println(plugin.getName());
+
+    var jniContract = KeySet.create();
+    try {
+      plugin.get(jniContract, Key.create(Plugin.JNI_MODULE_CONTRACT_ROOT));
+    } catch (KDBException e) {
+      // ignored
+    }
+
+    var contract = KeySet.create();
+    for (Key key : jniContract) {
+      Key k = key.dup();
+      k.setName(
+          key.getName()
+              .replace(Plugin.JNI_MODULE_CONTRACT_ROOT, "system:/elektra/modules/stdioproc"));
+      contract.append(k);
+    }
+
+    contract.append(Key.create("system:/elektra/modules/stdioproc/exports/open", "1"));
+    contract.append(Key.create("system:/elektra/modules/stdioproc/exports/get", "1"));
+    contract.append(Key.create("system:/elektra/modules/stdioproc/exports/set", "1"));
+    contract.append(Key.create("system:/elektra/modules/stdioproc/exports/close", "1"));
```
I really don't like this bit. Adding at least a `getContract()` would be much better (even better would be also adding `getSupportedOperations()`).

kodebach replied:
It is/was planned with the new backend. Implementing the change separately would be a lot of unnecessary overhead.

markus2330 commented:
> The new plugin is more general, than the old one.

Please remove the old plugin process and rename your plugin to be called "process".

> Apart from tests/docs the PR is now fully read for review.

Please add tests+docs! Please also add code how to use Java plugins. :sparkling_heart:

markus2330 commented on src/plugins/dump/dump.cpp:
```diff
@@ -528,6 +587,7 @@ int elektraDumpGet (ckdb::Plugin * handle, ckdb::KeySet * returned, ckdb::Key *
 	keyDel (root);
 	int errnosave = errno;
 
+	// TODO: document
```
TODO, is this needed for the new version of plugin process, too? (remove if not)

kodebach replied:
I will remove this with the removal of the old `process` and renaming of `stdioproc` (after PR reviews are done).

markus2330 commented on src/plugins/stdioproc/CMakeLists.txt:
```diff
@@ -0,0 +1,15 @@
+
+if (ADDTESTING_PHASE AND ENABLE_KDB_TESTING AND "jna" IN_LIST ADDED_BINDINGS)
+	add_msr_test_plugin ("stdioproc" ENVIRONMENT "BUILD_DIR=${CMAKE_BINARY_DIR}")
+endif ()
```
```suggestion
endif ()

```

markus2330 commented on src/plugins/stdioproc/README.md:
```diff
@@ -0,0 +1,201 @@
+
+## Usage
+
+Set the config key `app` and the arrays `args/#`, `env/#` to the path of an executable, the arguments that shall be passed and the environment variables to be set.
```
app? I would say `executable`

markus2330 commented on src/plugins/stdioproc/README.md:
````diff
@@ -0,0 +1,201 @@
+Set the config key `app` and the arrays `args/#`, `env/#` to the path of an executable, the arguments that shall be passed and the environment variables to be set.
+
+```
+kdb mount test.dump /tests/stdioproc stdioproc 'app=/usr/bin/pluginproc' 'args=#1' 'args/#0=--load-plugin' 'args/#1=myplugin'
````
Do we provide this pluginproc as example?

kodebach replied:
No, this is a purely hypothetical example. A real world example (Java) can be found below. I didn't want to use the `testapp.sh` as an example, since it doesn't really do what the child process is "supposed to". It only does things that make it convenient for testing.

markus2330 commented on src/plugins/stdioproc/README.md:
```diff
@@ -0,0 +1,201 @@
+It then `fork`s a new process and the child calls `execve` with the path from `app` as well as `argv` and` envp`.
+The child process is expected to listen to `stdin` and write to `stdout` according to the protocol described below.
+
+If communication can be established, the child-process will be kept running until `elektraStdprocioClose`.
```
```suggestion
If communication can be established, the child process will be kept running until `elektraStdprocioClose`.
```

markus2330 commented on src/plugins/stdioproc/README.md:
```diff
@@ -0,0 +1,201 @@
+To encode keysets and keys we use the format of the `dump` plugin.
+It is important to note that the `dump` plugin is instructed to write the full keynames (normally it removes the parent prefix).
+There is, however, an exception during initialization, which is described in the appropriate section.
+The child-process may also delegate directly to the `dump` plugin, or re-implement the encoding.
```
```suggestion
The child process may also delegate directly to the `dump` plugin, or reimplement the encoding.
```

markus2330 commented on src/plugins/stdioproc/README.md:
````diff
@@ -0,0 +1,201 @@
+> ```
+> Parent > Child
+>
+> HELLO WORLD
````
protocol version here?

kodebach replied:
This is just an example of how the protocol is described within the README. It is not part of the actual protocol.

markus2330 commented on src/plugins/stdioproc/README.md:
```diff
@@ -0,0 +1,201 @@
+This will be used to replace `stdioproc` in the contract of the plugin.
+The `[contract]` keyset must contain the `infos` keys, which will be used replace the ones at the top of this README and the ones in `stdioproc.c`.
+It must also contain `exports/_` keys for every operation that is implemented by the child.
+All keys in `[contract]` should be below `system:/elektra/modules/stdioproc`.
```
Maybe allow any `system:/elektra/modules/<modulename>`?

kodebach replied:
IMO this is only an option, if `<modulename>` must match the previously sent `(name)`. This would also create a new source of errors.

markus2330 commented on src/plugins/stdioproc/README.md:
````diff
@@ -0,0 +1,201 @@
+A child that implements all operations should include these keys (in addition to the relevant `infos` keys):
+
+```
+system:/elektra/modules/stdioproc/exports/open = 1
````
Its not a good idea to reuse the same name for something different. What about

```suggestion
system:/elektra/modules/stdioproc/exports/has/open = 1
```

markus2330 commented on src/plugins/stdioproc/README.md:
````diff
@@ -0,0 +1,201 @@
+[data]
+```
+
+Here `(result)` is one of `success`, `noupdate` and `error`, which correspond to `ELEKTRA_PLUGIN_STATUS_SUCCESS`, `ELEKTRA_PLUGIN_STATUS_NO_UPDATE` and `ELEKTRA_PLUGIN_STATUS_ERROR` respectively.
````
Why not send ELEKTRA_PLUGIN_STATUS_SUCCESS and so on?

kodebach replied:
Why is sending a much longer string any better? We already have a handshake, so the `ELEKTRA_PLUGIN_STATUS_` prefix doesn't serve any purpose here.

If you mean sending the integer values of the constants, then this would just make the protocol less human-readable. It wouldn't actually have any benefit, because you'd still need to decode the string e.g. `1` into the integer value.

markus2330 commented on src/plugins/stdioproc/README.md:
```diff
@@ -0,0 +1,201 @@
+
+(result)
+[parent]
+[data]
```
```suggestion
[returned]
```

markus2330 commented on src/plugins/stdioproc/README.md:
````diff
@@ -0,0 +1,201 @@
+```
+Parent > Child
+
+(opname)
````
Please be a bit close to normal protocol descriptions, e.g.:

```suggestion
(open|get|set|close)
```

markus2330 commented on src/plugins/stdioproc/README.md:
```diff
@@ -0,0 +1,201 @@
+
+The child process should now exit (and thereby close its ends of the `stdin`/`stdout` pipes).
+
+> **Note:** Under normal circumstances this only happens, when plugin is being closed.
```
```suggestion
> **Note:** Under normal circumstances this only happens, after the plugin was being closed.
```

kodebach replied:
No it actually happens **while** the plugin is being closed, i.e. the `close` command is sent in the middle of a `elektraPluginClose` on an instance of `stdioproc`.

markus2330 commented on src/plugins/stdioproc/README.md:
````diff
@@ -0,0 +1,201 @@
+
+```sh
+# mount the Whitelist Java Plugin via stdioproc
+sudo kdb mount config.file user:/tests/stdioproc dump stdioproc 'app=/usr/bin/java' 'args=#3' 'args/#0=-cp' "args/#1=$BUILD_DIR/src/bindings/jna/plugins/whitelist/build/libs/whitelist-$(kdb --version | sed -nE 's/KDB_VERSION: (.+)/\1/gp')-all.jar" 'args/#2=org.libelektra.stdioproc.StdIoProcApp' 'args/#3=org.libelektra.plugin.WhitelistPlugin'
````
Can you please write a `mount-java` wrapper that takes the ugliness out of this?

kodebach replied:
In theory we could do something like

```sh
sudo kdb mount-java config.file user:/tests/stdioproc kdb:dump java:org.libelektra.plugin.WhitelistPlugin 
```

The `java:` and `kdb:` prefixes are to differentiate between `Java` plugins and others. This would also only work for Java plugins that are part of this repo. If you want to use plugins from other JARs, you'd have to define the `CLASSPATH` environment variable accordingly.

However, I don't think this example should be changed at all (adding `mount-java` is still a good idea). This example is not supposed to show how Java plugins are mounted. It is supposed to be an example of how to use `stdioproc`. Therefore, it should not hide the actual `stdioproc` config in some helper script.

markus2330 replied:
@kodebach wrote:
> In theory we could do something like
> 
> sudo kdb mount-java config.file user:/tests/stdioproc kdb:dump java:org.libelektra.plugin.WhitelistPlugin

I think we can assume `kdb:` if no `java:` is given?

> The |java:| and |kdb:| prefixes are to differentiate between |Java| 
> plugins and others. This would also only work for Java plugins that are 
> part of this repo. If you want to use plugins from other JARs, you'd 
> have to define the |CLASSPATH| environment variable accordingly.
> 
> However, I don't think this example should be changed at all (adding 
> |mount-java| is still a good idea). This example is not supposed to show 
> how Java plugins are mounted. It is supposed to be an example of how to 
> use |stdioproc|. Therefore, it should not hide the actual |stdioproc| 
> config in some helper script.

I agree, in the process plugin there should be the full example without any helper scripts. Could you please add/extend a Java-Plugin Tutorial where you describe `mount-java` and also how to set CLASSPATH for out-of-source-tree Java Plugins?

kodebach replied:
> I think we can assume kdb: if no java: is given?

I don't think there is technical reason why `java:com.example.Foo` couldn't be the name of a plugin written in C. The `.so` file certainly could contain `.` and `:`. But we could make `kdb:` optional. Then we'd have:
- `kdb:java:com.example.Foo` means C plugin named `java:com.example.Foo`
- `java:com.example.Foo` means Java plugin implemented in the class `com.example.Foo`
- `foo` and `kdb:foo` both mean C plugin named `foo`


>  Could you please add/extend a Java-Plugin Tutorial where you describe mount-java and also how to set CLASSPATH for out-of-source-tree Java Plugins?

Will do

markus2330 commented on src/plugins/stdioproc/stdioproc.c:
```diff
@@ -0,0 +1,576 @@
+	} hasOp;
+} IoData;
+
+#define MSG_HANDSHAKE_HEADER_V1 "ELEKTRA_STDIOPROC INIT v1\n"
```
const char *?

kodebach replied:
Then we can't use `sizeof`. In most other cases `strlen` would work and the compiler would optimize it away. But in the snippet below, it would change the meaning of the code.

https://github.com/ElektraInitiative/libelektra/blob/38d1d1789759aa8968575ad857e38f90d9c8c9ae/src/plugins/stdioproc/stdioproc.c#L148

markus2330 commented on src/plugins/stdioproc/stdioproc.c:
```diff
@@ -0,0 +1,576 @@
+		return ELEKTRA_PLUGIN_STATUS_SUCCESS;
+	}
+
+	Key * appKey = ksLookupByName (config, "/app", 0);
```
See above, `/executable`

markus2330 commented on src/plugins/stdioproc/README.md:
````diff
@@ -0,0 +1,201 @@
+```
+
+## Limitations
+
````
```suggestion

- The executable must have an absolute path.
```

markus2330 commented on src/plugins/stdioproc/stdioproc.c:
```diff
@@ -0,0 +1,576 @@
+
+	if (pipe (parentToChild) != 0)
+	{
+		ELEKTRA_SET_RESOURCE_ERRORF (errorKey, "Could not execute app. Reason: %s", strerror (errno));
```
Add more information:

- what the executable was
- that pipe failed

etc.

kodebach commented:
> Please add tests+docs! Please also add code how to use Java plugins. 💖

I have now added tests/docs. The README for `stdioproc` includes a Java plugin example.

> Please remove the old plugin process and rename your plugin to be called "process".

Will do before the PR is merged. I'll wait for your review, however, because renaming stuff mid-review breaks GitHub's comments.

markus2330 commented on src/plugins/stdioproc/testmod_stdioproc.c:
```diff
@@ -0,0 +1,180 @@
+
+#include <tests_plugin.h>
+
+// TODO: check contract
```
TODO

markus2330 reviewed (commented):
Great work! :rocket:

kodebach commented on src/bindings/jna/README.md:
````diff
@@ -49,7 +49,7 @@ export CLASSPATH=".:/usr/share/java/libelektra.jar:/usr/share/java/jna.jar"
 
 ```sh
-javac HelloElektra.java && java HelloElektra
+javac org.libelektra.stdioproc.StdIoProcApp.java && java HelloElektra
````
This was not an intentional change.

kodebach commented on src/plugins/dump/dump.cpp:
```diff
@@ -476,6 +476,63 @@ int unserialise (std::istream & is, ckdb::Key * parentKey, ckdb::KeySet * ks, bo
 	return ELEKTRA_PLUGIN_STATUS_SUCCESS;
 }
 
+class FileStreamBuf : public std::streambuf
```
I'll add comments, but basically it's the only way I found of reading from a `FILE *` via C++ streams.

kodebach commented:
@markus2330 I have now update the docs and added your other suggestions. Apart from the removal of the old `process` and renaming of `stdioproc` to `process`, this PR should be done IMO.

markus2330 reviewed (commented):
I reviewed already, please do last changes (removal+renaming), making buildserver happy, etc.

kodebach commented:
Everything is now renamed. I didn't remove the `pluginprocess` library, because it is also used by the `python` plugin. We should probably add a separate PR that removes the `python` plugin and creates helper scripts in Python similar to what we added here for Java.

markus2330 commented:
There are still several problems with the build server. Can you please fix/rerun?

Btw. did you check that everything gets installed? I'll test when merged with the packages.

kodebach commented:
> Can you please fix/rerun?

I'm not really sure, what's wrong. It works on my system, so I'll have to investigate a bit. From the error message it seems like Java can't find the `.so` files, which is weird because `LD_LIBRARY_PATH` should be set correctly.

> Btw. did you check that everything gets installed?

At least on my system everything gets installed, yes. I checked, because that's the only way to test `kdb mount-java` (at least the part that automatically sets the classpath).

kodebach commented:
I found the bug (`LD_LIBRARY_PATH` wasn't propagated to the child process), hopefully everything runs correctly now

markus2330 commented on src/plugins/dump/dump.cpp:
```diff
@@ -476,6 +476,68 @@ int unserialise (std::istream & is, ckdb::Key * parentKey, ckdb::KeySet * ks, bo
+	char buf_;
+};
+
+int freadks (KeySet * ks, FILE * file, Key * errorKey)
```
rename to fserialize, funserialize (also rename serialise and unserialise)

kodebach replied:
> also rename serialise and unserialise

Why?

markus2330 replied:
Because it is still British English and we decided to go for American English.

kodebach replied:
Oh okay, I thought you didn't like the term serialize/unserialise for some reason

markus2330 reviewed (commented):
Looks good to me to start testing, added some suggestions for better names.

kodebach commented on doc/tutorials/java-plugins.md:
```diff
@@ -11,42 +11,65 @@ If you need a tutorial for using Key and KeySet in Java, please check out the [H
+   Methods that are not supported, should simply return `Plugin.STATUS_SUCCESS`.
+2. In C the parent key of the contract depends on the plugins name.
+   For example, the contract for `dump` can be found under `system:/elektra/modules/dump` and the `dump` plugin returns it as such.
+   However, in Java the parent key for the contract is always `system:/elektra/modules/jni` (you may use the constant `Plugin.JNI_MODULE_CONTRACT_ROOT`).
```
I just kept the old key name to minimize the number of changes.

kodebach commented on doc/tutorials/java-plugins.md:
```diff
@@ -11,42 +11,65 @@ If you need a tutorial for using Key and KeySet in Java, please check out the [H
+1. In C it is possible to implement only `elektraPluginGet` and leave the other functions unimplemented.
+   Because of interface inheritance in Java, it is required to implement all 5 method `open`, `get`, `set`, `error` and `close`.
+   Whether a method is actually supported, must be communicated via the contract.
+   Methods that are not supported, should simply return `Plugin.STATUS_SUCCESS`.
```
A method that is not marked in the `exports/has` part of the contract shouldn't be called. However, returning `Plugin.STATUS_SUCCESS` is still safer, because if the method is somehow called, simply nothing happens.

Also, if you want to throw something, the appropriate type would be [`UnsupportedOperationException`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/UnsupportedOperationException.html). A `NoSuchMethodError` is a subtype of `Error` and therefore should not be caught and simply lead to a program crash.

markus2330 replied:
> because if the method is somehow called, simply nothing happens.

Why would it be "somehow" called? If export/has = 0, we could guarantee that we do not call it? And then it would not matter how the implementation is. A typical implementation in Java would be to throw `NoSuchMethodError`, as calling such a method is basically an assertion error (which also extends `Error`).

The sentences as they are written now sound like we might call it, thus the user must implement it as `Plugin.STATUS_SUCCESS`. This is problematic from maintenance point of view, as someone might think the implementation is as it should be while in fact it is not implemented.

kodebach replied:
> A typical implementation in Java would be to throw NoSuchMethodError, as calling such a method is basically an assertion error (which also extends Error).

Every implementation I've seen would use `UnsupportedOperationException` in this case. AFAIK `NoSuchMethodError` is basically only used by the JVM itself when bytecode for classes is modified at runtime or when using reflection.

---

Thinking about it, the best solution might actually be to use `default` implementations in the interface:

```java
public interface Plugin {
  default int open(KeySet config, Key errorKey) {
    throw new UnsupportedOperationException();
  }
  
  // no default because it must be implemented for the contract
  int get(KeySet keySet, Key parentKey) throws KDBException;

  default int set(KeySet keySet, Key parentKey) {
    throw new UnsupportedOperationException();
  }

  default int error(KeySet keySet, Key parentKey) {
    throw new UnsupportedOperationException();
  }

  default int close(Key parentKey) {
    throw new UnsupportedOperationException();
  }
}
```

kodebach commented on doc/tutorials/mount-java.md:
````diff
@@ -0,0 +1,33 @@
+```
+
+Here we added `/path/to/foo.jar` and `/path/to/bar.jar` to the classpath, so that the classes `org.example.Foo` and `org.example.Bar` can be found.
+We can still use `org.libelektra.plugin.WhitelistPlugin`, because Elektra's JARs are always added to the classpath.
````
Compiling the plugin isn't any different from compiling any other Java library into a JAR file. This is not supposed to be a Java programming tutorial, so I only detailed the parts that are special about Elektra plugins.

markus2330 replied:
As said: then simply put the information to the man page.

kodebach commented on src/bindings/jna/process/src/main/java/org/libelektra/process/ProcessApp.java:
```diff
@@ -0,0 +1,55 @@
+
+    try {
+      if (!protocol.handshake()) {
+        System.exit(1);
```
Since the handshake failed, the only real option here is to print something to `stderr`. However, that could be really confusing, because (under normal circumstances) we are not in the process that was called by the user.

Even if we print error messages to `stderr` (which could be really out of place depending on what the parent process is doing), in this specific case, I don't think we need one. If the handshake failed, there isn't much information we could provide. If the app was called by the `process` plugin and something went wrong, `process` will issue a "handshake failed" error. If the app was called in some other way, just exiting because the app is being misused seems fine.

markus2330 replied:
We could exit with 3 different error status? Then the plugin can wait for the child and knows the exit status for an error message.

kodebach replied:
I'm not sure that would have much benefit. The three cases would be:
1. "Handshake failed": Without any extra information the separate exit code is useless. The plugin already knows that the handshake failed, because it initiated the handshake which the child process aborted.
2. An IOException: In this case, maybe a different error code could convey some information. The parent process probably also got some error from `read`/`write`ing to its end of the pipes.
3. Another error during the protocol execution: This mostly boils down to an encoding/decoding error of the dump format. This shouldn't happen and if it does it's a bug.

Additionally, in all 3 cases the user can't really do anything to solve the error. Not even the developer of the Java plugin could do anything here. All these errors are essentially bugs that need to be reported and fixed by us (*).

---

(*) I didn't use `ELEKTRA_SET_INTERNAL_ERROR` in the `process` plugin, because the errors are only internal errors, if the child app is also provided us (like with Java). If it's an external app, the developer of that app needs to fix the problem.

markus2330 replied:
> "Handshake failed": Without any extra information the separate exit code is useless. The plugin already knows that the handshake failed, because it initiated the handshake which the child process aborted.

It does not know if the child agrees about that the handshake failed. Even if this is very unlikely/impossible there is also not a good reason to use the same exit code?

> An IOException: In this case, maybe a different error code could convey some information. The parent process probably also got some error from read/writeing to its end of the pipes.

Good idea!

> Additionally, in all 3 cases the user can't really do anything to solve the error. Not even the developer of the Java plugin could do anything here. All these errors are essentially bugs that need to be reported and fixed by us (*).

Yes, so the error should be designed that it is helpful for us.

markus2330 replied:
> What do you mean "the child agrees"?

The child might have written the ACK to stdout but the parent process does not receive it (e.g. wrongly connected stdout).

> On a more technical note, it also requires quite extensive changes to the process plugin's code.

If it makes the code much more complicated then leave it. We will see if errors in the protocol occurs and if we will be able to debug it easily :wink:

kodebach commented on src/bindings/jna/process/src/main/java/org/libelektra/process/ProcessApp.java:
```diff
@@ -0,0 +1,55 @@
+        System.exit(1);
+      }
+    } catch (IOException e) {
+      System.exit(1);
```
Where would this error message go? The `IOException` means that something went wrong while communicating with the parent process. Sending an error message via the protocol is clearly not an option here.

For more about errors see above.
